File: 백준/Gold/1753. 최단경로/최단경로.py
import sys
import heapq

input = sys.stdin.readline
hpush = heapq.heappush
hpop = heapq.heappop

v, e = map(int, input().split())
k = int(input())

graph = [dict() for _ in range(20001)]

for _ in range(e):
    _u, _v, _w = map(int, input().split())
    if graph[_u].get(_v) is None: 
        graph[_u][_v] = _w
    else:
        graph[_u][_v] = min(graph[_u][_v], _w)

INF = 99999999
dist = [INF] * 20001

dist[k] = 0
pq = [(0, k)]

while (pq):
    cost, cur = hpop(pq)
    for nxt_node, nxt_cost in graph[cur].items():
        nxt_cost += cost
        if (dist[nxt_node] < nxt_cost):
            continue
        dist[nxt_node] = nxt_cost
        hpush(pq, (nxt_cost, nxt_node))

for i in range(1, v + 1):
    if dist[i] == INF:
        print("INF")
    else:
        print(dist[i])

File: 백준/Gold/1753. 최단경로/최단경로.rs
use std::io::{stdin, stdout, Read, Write, BufReader, BufWriter};
use std::fs::File;
use std::collections::{BinaryHeap, HashMap};
use std::cmp::*;

fn main() {
    let mut reader = BufReader::new(stdin());
    
    let mut input = String::new();
    reader.read_to_string(&mut input).unwrap();
    let mut input = input.split_ascii_whitespace().flat_map(str::parse::<usize>);

    let (v, e) = (input.next().unwrap(), input.next().unwrap());
    let k = input.next().unwrap();

    let mut graph = vec![HashMap::<usize, usize>::new(); 20001];
    for _ in 0..e {
        let (_u, _v, _w) = (input.next().unwrap(), input.next().unwrap(), input.next().unwrap());
        graph[_u].entry(_v).and_modify(|e| {
            *e = min(*e, _w);
        }). or_insert(_w);
    }
    let max_dist = 99999999_usize;
    let mut dist = vec![max_dist; 20001];
    let mut pq = BinaryHeap::<Reverse<(usize, usize)>>::new();

    dist[k] = 0;
    pq.push(Reverse((0, k)));

    while let Some(Reverse((cost, cur))) = pq.pop() {
        if cost > dist[cur] {
            continue;
        }

        if let Some(connected) = graph.get(cur) {
            for (&nxt_node, &nxt_dist) in connected {
                let nxt_cost = nxt_dist + cost;
                if dist[nxt_node] > nxt_cost {
                    dist[nxt_node] = nxt_cost;
                    pq.push(Reverse((nxt_cost, nxt_node)));
                }
            }
        }
    }

    for i in 1..=v {
        if dist[i] == max_dist {
            println!("INF");
        } else {
            println!("{}", dist[i]);
        }
    }
}
